### ISSUE TITLE
Deallocating empty batches in ParticleData::compress to clean up memory leaks

### ISSUE THREAD

**Issue by stefanarridge** [enzo-project/enzo-e#142]
As mentioned in the description of PR #140 , there is a memory leak issue when running with the "merge_stars" method, if a too small batch size is used. This happens since the method involves copying all (star) particles on a given block to / from all neighbouring blocks, which can mean that multiple new batches have to be allocated to store all the temporary particles. After particles are deleted again, this leads to having (possibly multiple) empty batches. When particles are inserted / copied again, there are placed in the 'last batch' which causes even more batches to be allocated, and the problem is compounded. Eventually (in my experience) the code crashes when writing output.

I'm not sure there is an obvious way of preventing the memory leaks in the first place, other than by increasing the batch size in the parameter file, but at the very least, they could be cleaned up by compressing the particle data (as already implemented in ParticleData::compress), to ensure there are no 'empty spaces' interleaved with particle data in the batches, and then deallocating empty batches, which does not appear to be implemented yet, based on the comment at line 588 in src/Cello/data_ParticleData.cpp. This function could then be called every n timesteps (where n is a parameter for the user to set), or whenever the memory efficiency of the particle data drops below some threshold. 

@jobordner, what are your thoughts?

**Comment by jobordner** [comment]
Thanks, Sefan, I started looking into this again last week. There should never be empty spaces between particles because, when particles are deleted, the remaining particles in the batch are "pushed to the left" to fill the holes. But there can be (but shouldn't be) empty batches, which is a problem since the code assumes batches are non-empty. I think compress fails in that case.

**Comment by stefanarridge** [comment]
Thanks for the explanation James, just one thing I'd like to clarify: you wrote that "when particles are deleted, the remaining particles in the batch are "pushed to the left" to fill the holes". What does compress do in addition in that case? Does it move particles between batches to ensure that all batches up to some index are full?

**Comment by jobordner** [comment]
> Thanks for the explanation James, just one thing I'd like to clarify: you wrote that "when particles are deleted, the remaining particles in the batch are "pushed to the left" to fill the holes". What does compress do in addition in that case? Does it move particles between batches to ensure that all batches up to some index are full?

Yes, after compress, the only non-full batch should be the last one.

**Comment by stefanarridge** [comment]
> Yes, after compress, the only non-full batch should be the last one.

So are the empty batches deallocated?
